### PR TITLE
Added c++ coding best practices based on our code

### PR DESCRIPTION
**Pull request by dmakoviichuk-tt**
### Problem description
Improve quality and perfromance of the c++ code.
Document is a very small. All problems were found in our code. 
It is high priority to fix it.

### What's changed
Added DM doc. In future main README.dm will be have link on top of itself.

### Checklist
- [ ] Post commit CI passes
- [ ] Model regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: best_practices.md [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+### Explanation
+Using `const` rvalue references (`const T&&` or `const auto&&`) and returning `const` values from functions can prevent the compiler from optimizing the code. These practices block the use of move constructors, which are crucial for efficient resource management and performance optimization.
+
+### Motivation
+- **Performance**: Enables the use of move constructors, enhancing performance.
+- **Optimization**: Allows the compiler to optimize code more effectively.
+- **Efficiency**: Facilitates efficient resource management by leveraging move semantics.
+
+  ## 8. Use `std::move` with `emplace_back` and Prefer `push_back` When Appropriate
```
Comment (ayerofieiev-tt) [inline best_practices.md:93]:
```suggestion
## 8. Use `std::move` with `emplace_back` and Prefer `push_back` When Appropriate
```

File: best_practices.md [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+### Motivation
+- **Efficiency**: Ensures resources are moved, not copied, maintaining performance.
+- **Clarity**: Using `push_back` can be simpler and equally efficient in certain scenarios.
+- **Optimization**: Maximizes the benefits of in-place construction and move semantics.
+
+## 9. Move constructor should not throw. Mark Move Constructors as `noexcept`
+
+### Practice
+Move constructors should be marked as `noexcept`, otherwise many STL optimizations will not be applied.
```
Comment (ayerofieiev-tt) [inline best_practices.md:109]: This one is really important and too easy to overlook. Do you know if clang-tidy has any rule to catch this?

**Review by ayerofieiev-tt — APPROVED** [review]
I'd get it in like this. And add code snippets with another PR

File: best_practices.md (hunk 2 above)
Comment (dmakoviichuk-tt) [reply best_practices.md:109]: clang tidy is the next step 100%.

File: best_practices.md [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+
+### Explanation
+Global classes can lead to issues with concurrency and resource management, especially when they involve mutexes or other synchronization mechanisms. Such global resources can create hidden dependencies and make the code harder to reason about and maintain.
+
+### Motivation
+- **Concurrency Safety**: Reduces the risk of deadlocks and race conditions.
+- **Maintainability**: Avoids hidden dependencies and makes the code easier to understand and maintain.
+- **Predictability**: Improves the predictability of resource management and execution flow.
+
+## 12. Move Function Implementations to `.cpp` Files
```
Comment (dmakoviichuk-tt) [inline best_practices.md:237]: @ayerofieiev-tt  its here :)

File: best_practices.md [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+`std::span` is a lightweight view over a contiguous sequence of objects, such as arrays and vectors. It provides a safe and flexible way to handle array-like data structures without copying them.
+
+### Motivation
+- **Flexibility**: Enables functions to accept both `std::vector` and `std::array`.
+- **Efficiency**: Eliminates the need for copying data, enhancing performance.
+- **Safety**: Provides bounds-checked access to the underlying data.
+### Example
+```
+template <typename T>
+void print_elements(std::span<T> data) {
```
Comment (ayerofieiev-tt) [inline best_practices.md:36]: no const &?

File: best_practices.md [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+
+### Motivation
+- **User Experience**: Provides more informative error messages, improving user experience.
+- **Debugging**: Easier to debug issues with clear assertions and fatal errors.
+- **Performance**: Avoids the overhead associated with exception handling.
+
+### Example
+```
+for (int i = 0; i < fp32_vec.size(); i++) {
+  fp32_vec.at(i) = rand_float() + offset; // Very bad, breaks a lot of optimiations. Additional check.
```
Comment (ayerofieiev-tt) [inline best_practices.md:145]: Can we get a better example? we rarely use at to assign a value like this and I afraid that it does not land the point well.
I'd rather see an example where we just read with .at vs []

File: best_practices.md [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+RVO is an optimization technique that eliminates unnecessary copying or moving of objects when they are returned from a function. Using `std::move` can prevent the compiler from applying RVO, resulting in less efficient code.
+
+### Motivation
+- **Performance**: Ensures that RVO can be applied, minimizing unnecessary copying or moving of objects.
+- **Efficiency**: Maintains optimal performance by allowing the compiler to perform return value optimizations.
+
+### Example
+https://godbolt.org/z/enTqadjMz
+
+## 7. Avoid `const T&&` or `const auto&&` and Returning `const` Values
```
Comment (ayerofieiev-tt) [inline best_practices.md:167]: Wdyt about splitting these?

File: best_practices.md [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+
+### Explanation
+Implementing functions in `.cpp` files rather than in headers reduces compilation dependencies and improves encapsulation. It also leads to faster compile times and reduces the chances of multiple definition errors.
+
+### Motivation
+- **Compilation Time**: Reduces compilation times by minimizing dependencies.
+- **Encapsulation**: Improves encapsulation by hiding implementation details.
+- **Maintainability**: Simplifies code maintenance by separating interface from implementation.
+
+## 13. Avoid `using namespace std;` in Headers
```
Comment (ayerofieiev-tt) [inline best_practices.md:250]: Maybe "Avoid `using namespace` in headers"?

File: best_practices.md (hunk 4 above)
Comment (dmakoviichuk-tt) [reply best_practices.md:36]: Don't need it, 0 difference.

File: best_practices.md (hunk 5 above)
Comment (dmakoviichuk-tt) [reply best_practices.md:145]: It is example from our codebase. But Will add one more.

File: best_practices.md (hunk 6 above)
Comment (dmakoviichuk-tt) [reply best_practices.md:167]: It is about overusing const.  But you are riggt const T&& and return const are different problems.

File: best_practices.md (hunk 7 above)
Comment (dmakoviichuk-tt) [reply best_practices.md:250]: But I will mention that using namespace std is the worst.

File: best_practices.md (hunk 4 above)
Comment (dmakoviichuk-tt) [reply best_practices.md:36]: Even worse: https://quuxplusone.github.io/blog/2021/11/09/pass-string-view-by-value/

Comment (ayerofieiev-tt) [reply best_practices.md:36]: I was aware for string_view, but not about span. Good to know.
This slightly contradicts the "best practice" to pass by const &. Just want us to be mindful of that.

File: best_practices.md [hunk 8]
```diff
@@ -0,0 +1,263 @@
+# Best Practices for C++20 Repository
+
+## 1. Pass Complex Types by Const References
+
+### Practice
+Always pass all complex types (e.g., vectors, tensors) by const references.
```
[the remark this answers is not in the document]
Comment (dmakoviichuk-tt) [reply best_practices.md:6]: @KalaivaniMCW currently I saw optional tensors were passed by value which leads to copy.
In this case it should be reference or const reference.

**Pull request merged**